sdks/python/apache_beam/runners/portability/fn_api_runner/translations.py

```diff
@@ -690,6 +692,196 @@ def fix_side_input_pcoll_coders(stages, pipeline_context):
   return stages
 
 
+def pack_combiners(stages, context):
+  # type: (Iterable[Stage], TransformContext) -> Iterator[Stage]
+  """Packs sibling CombinePerKey stages into a single CombinePerKey.
+
+  If CombinePerKey stages have a common input, one input each, and one output
+  each, pack the stages into a single stage that runs all CombinePerKeys and
+  outputs resulting tuples to a new PCollection. A subsequent stage unpacks
+  tuples from this PCollection and sends them to the original output
+  PCollections.
+  """
+
+  class _UnpackFn(core.DoFn):
+    """A DoFn that unpacks a packed to multiple tagged outputs.
+
+    Example:
+      tags = (T1, T2, ...)
+      input = (K, (V1, V2, ...))
+      output = TaggedOutput(T1, (K, V1)), TaggedOutput(T2, (K, V1)), ...
+    """
+
+    def __init__(self, tags):
+      self._tags = tags
+
+    def process(self, element):
+      key, values = element
+      return [
+          core.pvalue.TaggedOutput(tag, (key, value))
+          for tag, value in zip(self._tags, values)
+      ]
+
+  def _get_fallback_coder_id():
+    return context.add_or_get_coder_id(
+        coders.registry.get_coder(object).to_runner_api(None))
+
+  def _get_component_coder_id_from_kv_coder(coder, index):
+    assert index < 2
+    if coder.spec.urn == common_urns.coders.KV.urn and len(
+        coder.component_coder_ids) == 2:
+      return coder.component_coder_ids[index]
+    return _get_fallback_coder_id()
+
+  def _get_key_coder_id_from_kv_coder(coder):
+    return _get_component_coder_id_from_kv_coder(coder, 0)
+
+  def _get_value_coder_id_from_kv_coder(coder):
+    return _get_component_coder_id_from_kv_coder(coder, 1)
+
+  def _try_fuse_stages(a, b):
+    if a.can_fuse(b, context):
+      return a.fuse(b)
+    else:
+      raise ValueError
+
+  # Group stages by parent, yielding ineligible stages.
+  combine_stages_by_input_pcoll_id = collections.defaultdict(list)
+  for stage in stages:
+    transform = only_transform(stage.transforms)
+    if transform.spec.urn == common_urns.composites.COMBINE_PER_KEY.urn and len(
```

Comment:
We need to group and filter on the environment of the transform as well, lest we try to pack Java or Go combiners here. 

The easiest way to do that would be to add a "beam:combinefn:packed_python:v1" URN at https://github.com/apache/beam/blob/master/sdks/python/apache_beam/portability/python_urns.py , inject it into the list at https://github.com/apache/beam/blob/master/sdks/python/apache_beam/transforms/environments.py#L587 , and then here you can look up the environment attached to the transform, and whether that URN shows up in its capabilities, so see if this is OK. 

I'd be glad to explain more if this doesn't make sense.

Reply:
Oh, I see you are checking for PICKLED_COMBINE_FN below. We should keep this, but the reason it may be insufficient is that one might use PICKLED_COMBINE_FN across different environments. (People using Python 2 transforms of an old version of Beam in Python 3 pipelines is an actual scenario that may come up soon as we're dropping Python 2 support.)

Reply:
OK. Keeping the check, but moving it further up.

Reply:
Also added a new test to make sure that we don't run combiner packing without the capability.